### PR TITLE
BREAKING: add support for k8s `disable-node-collector` flag

### LINKED ISSUE #6194: trivy k8s: `--disable-node-collector` flag
it is require to introduce a new flag `--disable-node-collector`
when the flag is in use trivy will not run `node-collector` job and the scan result will not present misconfiguration finding from inside the Node

- when the flag is not in use the `node-collector` job will not be executed and the scan result will not present misconfiguration finding from inside the Node

- remove old `--components` flag as it cannot operate in conjunction with this flag

### PR DESCRIPTION
**Pull request by chen-keinan**
## Description
Add support for k8s intrusive flag

## Related issues
- Close #6194

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).

### CODE REVIEW

**Review by knqyf263 — COMMENTED** [review]
It's a breaking change, right? We need a `BREAKING` prefix.
https://github.com/aquasecurity/trivy/blob/df024e88ddccc0bd9158e7a4a553983438399826/.github/workflows/semantic-pr.yaml#L31

Also, the integration test is failing.

**Comment by chen-keinan** [comment]
@knqyf263 @itaysk is `--non-intrusive` is more appropriate than `--skip-intrusive` as a flag name ?

**Comment by knqyf263** [comment]
First of all, I don't mind. @itaysk and you can decide it. I'm just thinking out loud. If the default mode is intrusive, should we consider a more understated name for the non-intrusive mode (such as `--passive` or `--unobtrusive`) rather than `--non-intrusive`? 

In some cases, negating a well-known or intuitive term can effectively convey the intended meaning. If "intrusive" is a term that users are likely to understand in the context of Trivy, then using "non-intrusive" as the flag name might be the most straightforward way to communicate the opposite behavior.

**Comment by itaysk** [comment]
In my opinion, `--intrusive`/`--non-intrusive` is fine. even though it doesn't convey the actual benefit for the user (like `node-collect` or `deep` or something like that), it gives the more important decision factor which is that this will actually create a job in the scanned cluster. I think generally a scanner is expected to be read-only, and manipulating the scan target is unconventional, and the user should understand the consequence.
This makes me think that perhaps it's even better to default to non-intrusive, and make the user explicitly say `--intrusive` in order to deploy the job, because there's inherent friction and risk in this action (as users reported). Even if we want and recommend that most users will do intrusive scan (and all our docs will include it by default), this puts the responsibility on the user for initiating the intrusive action rather than the tool to automatically do so. Akin to some tools requiring you to add `--yes` in order to perform the action. 

TL;DR:
I think `--intrusive/--non-intrusive` is the best terminology, and I'm leaning towards non-intrusive by default (but not fixed on this)

**Comment by itaysk** [comment]
after discussing this offline with @chen-keinan we agreed to not change the current default behavior, meaning by default the scan will be intrusive and the user can set a `--non-intrusive` flag to opt out.

**Comment by knqyf263** [comment]
I know the context well and it is difficult to get an impartial view of how people feel about current behavior. I'd request a quick review from @simar7.

**Review by knqyf263 — APPROVED** [review]
Thanks @simar7. Feedback from someone unfamiliar with k8s scanning is valuable.

On the other hand, we may be able to make the assumption that people who use the k8s subcommand should be somewhat familiar with it.

We can also take the view that it is okay because the CLI help provides supplementary explanations.

```
      --non-intrusive                     When the flag is activated, the node-collector job will not be executed, thus skipping misconfiguration findings on the node.
```

The final decision should be made by the code owner, Chen, so I'll approve it.

**Comment by itaysk** [comment]
I agree with simar and wrote similar gist on the --deep PR. `node-collector` works too IMO

**Comment by chen-keinan** [comment]
> > I know the context well and it is difficult to get an impartial view of how people feel about current behavior. I'd request a quick review from @simar7.
> 
> As I was summoned, I would like to provide my feedback.
> 
> I think there are two problems to solve here:
> 
> 1. Flag name
> 2. Flag default behavior
> 
> **Problem 1** I think personally, I would lean towards flags that are more "obvious" or "commonly used". To me "intrusive" (or lack thereof) doesn't tell me what's _really_ intrusive about it. To understand this I would have to read the Trivy docs to know the default behavior (or read the help for CLI flags).
> 
> While as something like `--node-collector-disable` (or something similar) conveys exactly what is being done, aka node collector scan results are being disabled. This also lines up well with existing `--node-collector-*` flags that we currently have within Trivy.
> 
> **Problem 2** If we change the flag name to something as I mentioned above, we can keep the current behavior as-is and have this flag opt-in. This would help us not make a breaking change.

@simar7 thanks for the input, putting a flag `--disable-node-collector` make it more understandable and specific for what actually this flag do.

File: pkg/k8s/report/report_test.go [hunk 1]
```diff
@@ -515,20 +515,15 @@ func Test_separateMisconfigReports(t *testing.T) {
 		name            string
 		k8sReport       Report
 		scanners        types.Scanners
-		components      []string
 		expectedReports []Report
 	}{
-		{
+		/*{
```
Comment (simar7) [inline pkg/k8s/report/report_test.go:520]: Did you forget to uncomment this test back in?

File: pkg/k8s/report/report_test.go [hunk 2]
```diff
@@ -567,37 +558,40 @@ func Test_separateMisconfigReports(t *testing.T) {
...
 				},
+				{
+					Resources: []Resource{
+						{Kind: "Pod"},
+					},
+				},
 			},
 		},
-		{
+		/*	{
```
Comment (simar7) [inline pkg/k8s/report/report_test.go:580]: ditto

**Comment by simar7** [comment]
@chen-keinan reviewed again and left one comment but overall lgtm!

**Comment by chen-keinan** [comment]
> @chen-keinan reviewed again and left one comment but overall lgtm!

@simar7 thanks for the review , I do not see the comment you mention

File: pkg/k8s/report/report_test.go (hunk 1 above)
Comment (simar7) [reply pkg/k8s/report/report_test.go]: It was this @chen-keinan looks like you resolved it.

**Comment by simar7** [comment]
> > @chen-keinan reviewed again and left one comment but overall lgtm!
> 
> 
> 
> @simar7 thanks for the review , I do not see the comment you mention

https://github.com/aquasecurity/trivy/pull/6311#discussion_r1580296070

**Review by simar7 — APPROVED** [review]

**Comment by simar7** [comment]
@chen-keinan lgtm!
